src/runtime/micro/micro_module.cc

```diff
@@ -0,0 +1,150 @@
+    session_ = MicroSession::Global();
+    low_level_device_ = session_->low_level_device();
+    binary_path_ = binary_path;
+    binary_info_ = session_->LoadBinary(binary_path_);
+
+    // Patch device lib pointers.
+    PatchImplHole("TVMBackendAllocWorkspace");
+    PatchImplHole("TVMBackendFreeWorkspace");
+    PatchImplHole("TVMAPISetLastError");
+  }
+
+  /*!
+   * \brief runs selected function on the micro device
+   * \param func_name name of the function to be run
+   * \param func_offset offset of the function to be run
+   * \param args type-erased arguments passed to the function
+   */
+  void RunFunction(const std::string& func_name, DevBaseOffset func_offset, const TVMArgs& args) {
+    session_->PushToExecQueue(func_offset, args);
+  }
+
+ private:
+  /*! \brief module binary info */
+  BinaryInfo binary_info_;
+  /*! \brief path to module binary */
+  std::string binary_path_;
+  /*! \brief global session pointer */
+  std::shared_ptr<MicroSession> session_;
+  /*! \brief low-level device pointer */
+  std::shared_ptr<LowLevelDevice> low_level_device_;
+
+  SymbolMap& symbol_map() {
+    return binary_info_.symbol_map;
+  }
+
+  /*!
+   * \brief patches a function pointer in this module to an implementation
+   * \param func_name name of the function pointer being patched
+   */
+  void PatchImplHole(const std::string& func_name) {
+    const DevBaseOffset init_impl_offset = session_->init_symbol_map()[func_name];
+    void* init_impl_addr = (low_level_device_->base_addr() + init_impl_offset).cast_to<void*>();
+    std::stringstream func_name_underscore;
+    func_name_underscore << func_name << "_";
+    const DevBaseOffset lib_hole_offset = symbol_map()[func_name_underscore.str()];
+    session_->DevSymbolWrite(symbol_map(), func_name_underscore.str(), init_impl_addr);
+  }
+};
+
+class MicroWrappedFunc {
+ public:
+  MicroWrappedFunc(MicroModuleNode* m,
+                   std::shared_ptr<MicroSession> session,
+                   const std::string& func_name,
+                   DevBaseOffset func_offset) {
+    m_ = m;
+    session_ = session;
+    func_name_ = func_name;
+    func_offset_ = func_offset;
+  }
+
+  void operator()(TVMArgs args, TVMRetValue* rv, void** void_args) const {
+    m_->RunFunction(func_name_, func_offset_, args);
+  }
+
+ private:
+  /*! \brief internal module */
+  MicroModuleNode* m_;
+  /*! \brief reference to the session for this function (to keep the session alive) */
+  std::shared_ptr<MicroSession> session_;
+  /*! \brief name of the function */
+  std::string func_name_;
+  /*! \brief offset of the function to be called */
+  DevBaseOffset func_offset_;
+};
+
+PackedFunc MicroModuleNode::GetFunction(
+    const std::string& name,
+    const std::shared_ptr<ModuleNode>& sptr_to_self) {
+  DevBaseOffset func_offset = symbol_map()[name];
+  MicroWrappedFunc f(this, this->session_, name, func_offset);
+  return PackFuncVoidAddr(f, std::vector<TVMType>());
```

Comment:
You do not need to use PackFunVoidAddr, this function packs the arguments into addresses that are needed by CUDA calling convention. You can directly create a PackedFunc that takes args and rv